src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/IconUrlToImageCacheConverter.cs

```diff
@@ -24,38 +29,127 @@ internal class IconUrlToImageCacheConverter : IValueConverter
+        private static readonly RequestCachePolicy RequestCacheIfAvailable = new RequestCachePolicy(RequestCacheLevel.CacheIfAvailable);
+
+        private static readonly ErrorFloodGate ErrorFloodGate = new ErrorFloodGate();
+
+        /// <summary>
+        /// Converts IconUrl from PackageItemListViewModel to an image represented by a BitmapSource
+        /// </summary>
+        /// <remarks>
+        /// We bind to a BitmapImage instead of a Uri so that we can control the decode size, since we are displaying 32x32 images, while many of the images are 128x128 or larger.
+        /// This leads to a memory savings.
+        /// </remarks>
+        /// <param name="values">
+        /// <list type="bullet">
+        /// <item>
+        /// <description>values[0]: IconUrl that points to a URL o a local file</description>
+        /// </item>
+        /// <item>
+        /// <description>values[1]: An <c>PackageArchiveReader</c> to read from the local package for embedded icons</description>
+        /// </item>
+        /// </list>
+        /// </param>
+        /// <param name="targetType">unused</param>
+        /// <param name="parameter">A BitmapImage with the default package icon</param>
+        /// <param name="culture">unused</param>
+        /// <returns>A BitmapSource with the image</returns>
+        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
         {
-            var iconUrl = value as Uri;
+            if (values == null || values.Length == 0)
+            {
+                return null;
+            }
+
+            var iconUrl = values[0] as Uri;
             var defaultPackageIcon = parameter as BitmapSource;
             if (iconUrl == null)
             {
                 return null;
             }
 
-            var cachedBitmapImage = _bitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
+            var cachedBitmapImage = BitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
             if (cachedBitmapImage != null)
             {
                 return cachedBitmapImage;
             }
 
             // Some people run on networks with internal NuGet feeds, but no access to the package images on the internet.
             // This is meant to detect that kind of case, and stop spamming the network, so the app remains responsive.
-            if (_errorFloodGate.IsOpen)
+            if (ErrorFloodGate.IsOpen)
             {
                 return defaultPackageIcon;
             }
 
             var iconBitmapImage = new BitmapImage();
             iconBitmapImage.BeginInit();
-            iconBitmapImage.UriSource = iconUrl;
 
+            BitmapSource imageResult;
+
+            // Check if the URI is an Embedded Icon URI
+            if (IsEmbeddedIconUri(iconUrl))
+            {
+                // Check if we have enough info to read the icon from the package
+                if (values.Length == 2 && values[1] is Lazy<PackageReaderBase>)
```

Comment:
When `values[1] is Lazy<PackageReaderBase>` is `Lazy<PackageReaderBase>.Value` guaranteed to return a non-disposed `PackageReaderBase` instance?

I see the `PackageReaderBase` instance being disposed down below.  If this code path is called again, is there a possibility that the disposed instance could be reused?

For instance, the image cache caches images for 10 minutes.  Suppose you open PM UI to display icons from local packages and the total number of viewable packages is significantly greater than the number of packages in the virtualizing stack panel.  Scroll through the entire list of packages to ensure all icons have been loaded and cached.  Wait for 15 minutes.  Now scroll through the entire list again.  Are there any exceptions or issues reloading icons again?

Reply:
> When `values[1] is Lazy<PackageReaderBase>` is `Lazy<PackageReaderBase>.Value` guaranteed to return a non-disposed `PackageReaderBase` instance?

No. In LocalPackageSearchMetadata, the PackageArchiveReader property is implemented as a singleton.

> I see the `PackageReaderBase` instance being disposed down below.  If this code path is called again, is there a possibility that the disposed instance could be reused?

Yes. because the PackageArchiveReader is never created again.

> For instance, the image cache caches images for 10 minutes.  Suppose you open PM UI to display icons from local packages and the total number of viewable packages is significantly greater than the number of packages in the virtualizing stack panel.  Scroll through the entire list of packages to ensure all icons have been loaded and cached.  Wait for 15 minutes.  Now scroll through the entire list again.  Are there any exceptions or issues reloading icons again?

Yes, see https://github.com/NuGet/Home/issues/9069

There's a PR for fixing that https://github.com/NuGet/NuGet.Client/pull/3192